core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java

```diff
@@ -0,0 +1,459 @@
+
+         // If phase is READ_ALL_WRITE_ALL, we know that ST has finished and that it is safe to retry postponed version requests
+         VersionRequest request;
+         while ((request = retryQueue.poll()) != null) {
+            if (trace) log.tracef("Retrying %s", request);
+            request.start();
+         }
+      }
+   }
+
+   @Override
+   public Map<Address, InternalCacheValue<V>> getAllVersions(final K key) {
+      final VersionRequest request;
+      synchronized (versionRequestMap) {
+         request = versionRequestMap.computeIfAbsent(key, k -> new VersionRequest(k, stateConsumer.isStateTransferInProgress()));
+      }
+
+      try {
+         return request.completableFuture.get();
+      } catch (InterruptedException e) {
+         Thread.currentThread().interrupt();
+         throw new CacheException(e);
+      } catch (ExecutionException e) {
+         if (e.getCause() instanceof CacheException)
+            throw (CacheException) e.getCause();
+
+         throw new CacheException(e.getCause());
+      } finally {
+         synchronized (versionRequestMap) {
+            versionRequestMap.remove(key);
+         }
+      }
+   }
+
+   @Override
+   public Stream<Map<Address, InternalCacheEntry<K, V>>> getConflicts() {
+      return getConflicts(installedTopology);
+   }
+
+   private Stream<Map<Address, InternalCacheEntry<K, V>>> getConflicts(LocalizedCacheTopology topology) {
+      if (stateConsumer.isStateTransferInProgress())
+         throw log.getConflictsStateTransferInProgress(cacheName);
+
+      if (!streamInProgress.compareAndSet(false, true))
+         throw log.getConflictsAlreadyInProgress();
+
+      try {
+         return StreamSupport
+               .stream(new ReplicaSpliterator(topology), false)
+               .filter(filterConsistentEntries());
+      } finally {
+         streamInProgress.compareAndSet(true, false);
+      }
+   }
+
+   @Override
+   public boolean isConflictResolutionInProgress() {
+      return streamInProgress.get();
+   }
+
+   @Override
+   public void resolveConflicts() {
+      resolveConflicts(installedTopology);
+   }
+
+   public void resolveConflicts(CacheTopology topology) {
+      LocalizedCacheTopology localizedTopology;
+      if (topology instanceof LocalizedCacheTopology) {
+         localizedTopology = (LocalizedCacheTopology) topology;
+      } else {
+         localizedTopology = createLocalizedTopology(topology);
+      }
+      doResolveConflicts(localizedTopology);
+   }
+
+   private void doResolveConflicts(final LocalizedCacheTopology topology) {
+      final Set<Address> preferredPartition = new HashSet<>(topology.getCurrentCH().getMembers());
+
+      if (trace) log.tracef("Attempting to resolve conflicts.  All Members %s, Installed topology %s, Preferred Partition %s",
+            topology.getMembers(), topology, preferredPartition);
+
+      List<CompletableFuture<V>> completableFutures =
```

Comment:
Uh, won't this store a CF for each entry, not releasing anything until the end of ST? I would suggest AtomicLong (initialized with 1 and decremented once you request data for all keys) + CountDownLatch instead, as you don't care about the return values anyway. Maybe you'd need a semaphore, too, to limit the parallelism to a reasonable degree.
Another solution (more streamish) would be to call a reduce phase that would drop completed futures automatically, but it might be more tricky to tweak the `Characteristics` to make sure that you won't keep the CFs around.